### PR TITLE
Better error for ART <Group />

### LINKED ISSUE #3815: Native exceptions when manipulating ART children without a surface
Rendering this component throws a cryptic error on the native side because every ART Group should be mounted in a Surface and it gets confused. 

```js
'use strict';

var React = require('react-native');
var ReactART = require('ReactNativeART');

var {View} = React;
var {Group} = ReactART;

var ArtBug = React.createClass({
  getInitialState: function() {
    return {showGroup: true};
  },
  componentDidMount: function() {
    this.setState({showGroup: false});
  },
  render: function() {
    return (
      <View>{this.state.showGroup && <Group />}</View>
    );
  },
});

module.exports = ArtBug;
```

We should probably throw an error earlier for this case. One way to do it would be to use context to pass down some flag from the Surface and throw if it's not present in the other components.

### PR DESCRIPTION
**Pull request by grabbou**
Fixes #3815

CC: @spicyj

### CODE REVIEW

File: Libraries/ART/ReactNativeART.js [hunk 1]
```diff
@@ -205,9 +211,15 @@ function extractOpacity(props) {
...
 var Group = React.createClass({
-
+  contextTypes: {
+    isInSurface: React.PropTypes.bool.isRequired,
+  },
   render: function() {
     var props = this.props;
+    invariant(
+      this.context.isInSurface,
+      'ART: <Group /> must be a child of a <Surface />'
```
Comment (janicduplessis) [inline Libraries/ART/ReactNativeART.js]: Is there enough context to easily find what component caused this error? Maybe invariant stack trace is enough.

**Comment by janicduplessis** [comment]
@grabbou LGTM, wanna rebase to get tests passing?

@spicyj Do you have anything to add on this since you opened the original issue?

**Comment by spicyj** [comment]
We should add a check for all the components, not just Group.

**Comment by facebook-github-bot-9** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @brentvatne as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by javache** [comment]
Could you rebase this?

**Comment by lacker** [comment]
Let's not let the perfect be the enemy of the good here. IMO if this just rebases to solve the merge conflicts then it's good to get in as is, even if it's just Group rather than everything. @grabbou thoughts?

File: Libraries/ART/ReactNativeART.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ var ReactNativeViewAttributes = require('ReactNativeViewAttributes');
 
 var createReactNativeComponentClass = require('createReactNativeComponentClass');
 var merge = require('merge');
+var invariant = require('invariant');
```
Comment (leeight) [inline Libraries/ART/ReactNativeART.js]: Should be `fbjs/lib/invariant`

**Comment by hramos** [comment]
@grabbou do you still want to ship this?

**Comment by grabbou** [comment]
Yes, it's on my list. Sorry for the delay, will do it first thing in the morning.

**Comment by lacker** [comment]
Ping @grabbou first thing in the morning eh 😛

**Comment by grabbou** [comment]
Haha, luckily I am playing around with a new task management system recently which means no more "Will do" that are to be pinged again.

Saved :)

**Comment by javache** [comment]
@facebook-github-bot shipit

**Comment by javache** [comment]
Commited in https://github.com/facebook/react-native/commit/6feffe1de0dea0ed7aaba2d945c2f06c56f957e6

**Pull request closed without merging**